File: cli-go/internal/proposal/edit_test.go
package proposal

import "testing"

// TestApplyEdit_NoOpSendsNothing is the subtle, load-bearing rule (s07 §T4): an
// edit session that changes nothing must produce NO editedPayload, so "approved
// clean" is never faked into "approved after edit".
func TestApplyEdit_NoOpSendsNothing(t *testing.T) {
	payload := map[string]any{
		"to":      "outside@example.com",
		"subject": "Re: the thread",
		"text":    "the agent's draft",
		"blobId":  "b_1",
	}

	// Reply edit that re-supplies the SAME subject/text → no-op.
	edited, problem := ApplyEdit(payload, EditorForm{
		Shape: "reply", Subject: "Re: the thread", Text: "the agent's draft",
	})
	if problem != "" {
		t.Fatalf("unexpected problem: %s", problem)
	}
	if edited != nil {
		t.Errorf("no-op reply edit produced editedPayload %v — must be nil", edited)
	}

	// A real change → editedPayload, with the send-shaped fields preserved.
	edited, problem = ApplyEdit(payload, EditorForm{
		Shape: "reply", Subject: "Re: the thread", Text: "a human rewrite",
	})
	if problem != "" || edited == nil {
		t.Fatalf("real edit: edited=%v problem=%q", edited, problem)
	}
	if edited["text"] != "a human rewrite" {
		t.Errorf("edited text = %v", edited["text"])
	}
	if edited["to"] != "outside@example.com" || edited["blobId"] != "b_1" {
		t.Errorf("edit must preserve to/blobId, got %v", edited)
	}
}

func TestApplyEdit_JSONShape(t *testing.T) {
	payload := map[string]any{"a": float64(1), "b": float64(2)}

	// Reformatting only (key order flipped) is NOT an edit.
	edited, problem := ApplyEdit(payload, EditorForm{Shape: "json", JSON: `{"b":2,"a":1}`})
	if problem != "" || edited != nil {
		t.Errorf("reformat-only json: edited=%v problem=%q — must be a no-op", edited, problem)
	}

	// A value change is an edit.
	edited, problem = ApplyEdit(payload, EditorForm{Shape: "json", JSON: `{"a":1,"b":3}`})
	if problem != "" || edited == nil || edited["b"] != float64(3) {
		t.Errorf("changed json: edited=%v problem=%q", edited, problem)
	}

	// Bad JSON and non-object JSON both refuse and send nothing.
	if _, p := ApplyEdit(payload, EditorForm{Shape: "json", JSON: `{oops`}); p == "" {
		t.Error("malformed json must report a problem")
	}
	if _, p := ApplyEdit(payload, EditorForm{Shape: "json", JSON: `[1,2,3]`}); p == "" {
		t.Error("a JSON array is not a payload object — must refuse")
	}
}

func TestSameJSON_KeyOrderIndependent(t *testing.T) {
	a := map[string]any{"x": float64(1), "y": []any{float64(1), float64(2)}}
	b := map[string]any{"y": []any{float64(1), float64(2)}, "x": float64(1)}
	if !SameJSON(a, b) {
		t.Error("SameJSON must ignore key order")
	}
	c := map[string]any{"x": float64(1), "y": []any{float64(2), float64(1)}} // array order matters
	if SameJSON(a, c) {
		t.Error("SameJSON must respect array order")
	}
}

func TestEditorFor(t *testing.T) {
	// grant-request is not editable — approve/deny as asked (arch.md §1).
	if _, ok := EditorFor("grant-request", map[string]any{}); ok {
		t.Error("grant-request must not be editable")
	}
	// reply-draft opens a reply form pre-filled from the payload.
	form, ok := EditorFor("reply-draft", map[string]any{"subject": "S", "text": "T"})
	if !ok || form.Shape != "reply" || form.Subject != "S" || form.Text != "T" {
		t.Errorf("reply-draft editor = %+v ok=%v", form, ok)
	}
	// anything else opens a whole-payload JSON editor.
	form, ok = EditorFor("organize-files", map[string]any{"moves": float64(3)})
	if !ok || form.Shape != "json" {
		t.Errorf("organize-files editor = %+v ok=%v", form, ok)
	}
}

func TestPayloadDiff(t *testing.T) {
	orig := map[string]any{"keep": float64(1), "change": "old", "drop": true}
	edit := map[string]any{"keep": float64(1), "change": "new", "add": "x"}
	diffs := PayloadDiff(orig, edit)

	byKey := map[string]FieldDiff{}
	for _, d := range diffs {
		byKey[d.Key] = d
	}
	if _, ok := byKey["keep"]; ok {
		t.Error("unchanged key must not appear in the diff")
	}
	if d := byKey["change"]; d.Before != "old" || d.After != "new" {
		t.Errorf("change diff = %+v", d)
	}
	if d := byKey["drop"]; !d.Removed {
		t.Errorf("dropped key must be marked Removed, got %+v", d)
	}
	if d := byKey["add"]; !d.Added || d.After != "x" {
		t.Errorf("added key must be marked Added, got %+v", d)
	}
}

File: webmail/src/lib/approvals/edit.test.ts
import { describe, expect, it } from "vitest";
import { applyEdit, diffLines, editorFor, payloadDiff } from "./edit";

const replyPayload = {
  to: "grace@example.test",
  self: "eric@bullmoose.test",
  subject: "Re: Project Elk kickoff",
  text: "Monday works.\nSee you then.\n— Eric",
  blobId: "blob-1",
  inReplyTo: "m1@x",
  messageId: "m2@x",
  mode: "send",
};

describe("editorFor — what Edit opens", () => {
  it("gives reply-shaped kinds a subject/text form seeded from the payload", () => {
    const form = editorFor({ kind: "reply-draft", payload: replyPayload });
    expect(form).toEqual({
      shape: "reply",
      subject: "Re: Project Elk kickoff",
      text: replyPayload.text,
    });
    expect(editorFor({ kind: "start-thread", payload: replyPayload })?.shape).toBe("reply");
  });

  it("gives other kinds the payload as JSON", () => {
    const form = editorFor({ kind: "create-event", payload: { title: "Standup" } });
    expect(form?.shape).toBe("json");
    expect(JSON.parse((form as { json: string }).json)).toEqual({ title: "Standup" });
  });

  it("refuses to edit a grant-request — approve or deny as asked", () => {
    expect(editorFor({ kind: "grant-request", payload: { scope: "read" } })).toBeNull();
  });
});

describe("applyEdit — the retained-diff contract", () => {
  it("merges the human's subject/text over the payload, leaving the send fields intact", () => {
    const { editedPayload, problem } = applyEdit(replyPayload, {
      shape: "reply",
      subject: "Re: Project Elk kickoff",
      text: "Monday works.\nLet's keep 30 min for scope.\n— Eric",
    });
    expect(problem).toBeUndefined();
    expect(editedPayload).toMatchObject({
      text: "Monday works.\nLet's keep 30 min for scope.\n— Eric",
      // Untouched fields survive: the server validates to/self/blobId on the
      // effective payload (actionProposal.ts:429-436), so losing them here
      // would turn every edited approval into a refusal.
      to: "grace@example.test",
      blobId: "blob-1",
      mode: "send",
    });
  });

  it("a NO-OP edit sends nothing — approved clean must stay approved clean", () => {
    // The three-outcome table (s07 §T4): "approved after edit" is its own
    // signal, and opening the editor without changing anything must not fake it.
    const untouched = applyEdit(replyPayload, {
      shape: "reply",
      subject: replyPayload.subject,
      text: replyPayload.text,
    });
    expect(untouched.editedPayload).toBeUndefined();
    expect(untouched.problem).toBeUndefined();
  });

  it("a JSON no-op (reordered keys) is still a no-op", () => {
    const { editedPayload } = applyEdit(
      { a: 1, b: { c: 2, d: 3 } },
      { shape: "json", json: JSON.stringify({ b: { d: 3, c: 2 }, a: 1 }) },
    );
    expect(editedPayload).toBeUndefined();
  });

  it("refuses bad JSON and non-object JSON without sending anything", () => {
    expect(applyEdit({}, { shape: "json", json: "{nope" }).problem).toContain("not valid JSON");
    // Mirrors the server's own guard: `editedPayload must be an object`.
    expect(applyEdit({}, { shape: "json", json: "[1,2]" }).problem).toContain(
      "must be a JSON object",
    );
    expect(applyEdit({}, { shape: "json", json: "null" }).problem).toContain(
      "must be a JSON object",
    );
  });
});

describe("payloadDiff — the agent's version vs the human's, both retained", () => {
  it("reports changed, removed and added keys; identical keys are silent", () => {
    const original = { to: "a@x", text: "old", mode: "send", extra: 1 };
    const edited = { to: "a@x", text: "new", mode: "send", added: true };
    expect(payloadDiff(original, edited)).toEqual([
      { key: "text", before: "old", after: "new" },
      { key: "extra", before: 1 },
      { key: "added", after: true },
    ]);
  });

  it("never mutates either side — the original is the retained artifact", () => {
    const original = { text: "old" };
    const edited = { text: "new" };
    payloadDiff(original, edited);
    expect(original).toEqual({ text: "old" });
    expect(edited).toEqual({ text: "new" });
  });
});

describe("diffLines", () => {
  it("marks a rewritten line as del+add and keeps the common lines", () => {
    expect(diffLines("one\ntwo\nthree", "one\n2\nthree")).toEqual([
      { op: "same", text: "one" },
      { op: "del", text: "two" },
      { op: "add", text: "2" },
      { op: "same", text: "three" },
    ]);
  });

  it("handles pure insertion and pure deletion at the ends", () => {
    expect(diffLines("a\nb", "a\nb\nc")).toEqual([
      { op: "same", text: "a" },
      { op: "same", text: "b" },
      { op: "add", text: "c" },
    ]);
    expect(diffLines("a\nb", "b")).toEqual([
      { op: "del", text: "a" },
      { op: "same", text: "b" },
    ]);
  });

  it("identical inputs diff to all-same", () => {
    expect(diffLines("x\ny", "x\ny").every((l) => l.op === "same")).toBe(true);
  });
});
